Sensor updates

6by9 opened the pull request:
Updates OV5647 fora load of required controls.
Minor fix to imx219.
Adds [G|S]_SELECTION support to Unicam.

6by9 committed:
media: i2c: imx219: Declare that the driver can create events

The flag V4L2_SUBDEV_FL_HAS_EVENTS is required if the subdev can
generate events. It can create events from the ctrl handler, therefore
this is required.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: ov5647: Add support for g_selection to reflect cropping/binning

In order to apply lens shading correctly the client needs to know how
each mode crops or scales the image compared to the full sensor array.
Implement this (based on the imx219 equivalent).

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: ov5467: Fixup error path to release mutex

"87f3ab9 media: ov5647: Add basic support for multiple sensor modes."
added a return path ov5647_set_fmt that didn't release the device
mutex that it had claimed.
Release the mutex.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: ov5647: Support V4L2_CID_PIXEL_RATE

Clients need to know the pixel rate in order to compute exposure
and frame rate values.
Advertise it.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: bcm2835-unicam: Add support for VIDIOC_[S|G]_SELECTION

Sensors are now reflecting cropping and scaling parameters through
the selection API, therefore Unicam needs to forward the requests
through to the subdev.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: ov5647: Set V4L2_SUBDEV_FL_HAS_EVENTS flag

The ov5647 subdev can generate control events, therefore set
the V4L2_SUBDEV_FL_HAS_EVENTS flag.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
@naushir Need CID_VBLANK still.

6by9 committed:
media: i2c: ov5647: Add support for V4L2_CID_VBLANK

Adds vblank control to allow for frame rate control.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: ov5647: Neither analogue gain nor exposure need EXECUTE_ON_WRITE

The controls for analogue gain and exposure were defined with
V4L2_CTRL_FLAG_EXECUTE_ON_WRITE. This is not required as we only need
to send changes to the sensor.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: bcm2835-unicam: Do not stop streaming in unicam_release

unicam_release calls _vb2_fop_release, which will call stop_streaming
if that particular node was streaming. Calling it unconditionally (as
the code was) means that if a second handle was opened eg to alter
a setting, on closing that connection it also stopped Unicam.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: ov5647: Use member names in mode tables

To make adding new members to the mode structures easier, use
the member names in the initialisers.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
OK, I think this may be near complete. Please test and see what you get.

6by9 commented:
Take it back. The exposure control range isn't being set to that defined by VBLANK - sorry but the orriginal patch adding exposure is just plain wrong.
Nearly there though.

6by9 committed:
media: bcm2835-unicam: Fix reference counting in unicam_open

The reference counting of node->open was only incremented after
a check that the node was v4l2_fh_is_singular_file, which resulted
in the counting going wrong and s_power not being called at an
appropriate time.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

naushir commented on drivers/media/i2c/ov5647.c:
```diff
@@ -613,65 +618,124 @@ static struct ov5647_mode supported_modes_10bit[] = {
+			.top = 0,
+			.width = 2624,
+			.height = 1956,
+		},
```
Should this not be 2592x1944?

6by9 replied:
This was the mess I got into over native array vs active vs TIMING_x_OUTPUT_SIZE vs TIMING_x_ADDR_END-TIMING_x_ADDR_START vs what the heck was wanted by userspace.

Yes, 2592x1944 is probably correct.

6by9 committed:
media: i2c: ov5647: Advertise the correct exposure range

Exposure is clipped by the VTS of the mode, so needs to be updated as
and when this is changed.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Updated. I think I'm done. Flips can wait for now.
@naushir If you can review and test, then say if you're happy.

naushir commented:
LGTM

davidplowman commented:
I've been testing this with my ov5647 and libcamera - looks good to me too.

naushir commented:
@pelwell @popcornmix, this is ready to be merged if you are happy with it.
